### DISCUSSION TITLE
how to calculate the distance to root AST node?

### DISCUSSION THREAD

**Discussion by cyrfer** [babel/babel#14459] [General]
Hi, I'm new to using Babel as plugin author.

I am parsing Elastic Search mappings and I want to generate a few DSLs from the ES mappings. The mappings are JSON, so I could avoid Babel, but the Babel AST provided extra context that seemed useful when transforming.

The parse succeeds, providing a rich AST, and I traverse with a visitor for `ObjectExpression` nodes, with some filtering that identifies the nodes I really care about. I use traversal as a tool to simply collect the array of ObjectExpression paths/nodes. Once I have the array of AST Nodes, I transform each into a separate DSL AST library to generate the output DSL.

My problem is the DSL generation is failing, and I suspect it is because the order of the DSL elements is wrong. I want to re-order, to make nested elements first, and the root should be last.

How can I order based on distance to root? Does the Babel AST API make it easy to count edges to the root?

Thanks for any help.

**Comment by JLHwung** [comment]
> I want to re-order, to make nested elements first, and the root should be last.

You can hook on the exit listener if you would like the visitor to be invoked from leaves to the root:

```js
traverse(ast, {
  exit(path) {
    // do something
  }
}
```

**Reply by cyrfer** [reply]
Hi @JLHwung,
Thank you for looking. 

I think you are saying I can avoid any need to sort by applying transformations upon `exit`. That's a great idea and I will implement it immediately.

> I use traversal as a tool to simply collect the array of ObjectExpression path/nodes

To be clear this is what I was doing,
```javascript
  const esTypeNodes: Array< NodePath<ObjectExpression> > = []

  class ElasticTypeCollector implements TraverseOptions {
    enter(path: NodePath<Node>) {
      if (path.isObjectExpression()) {
        const pathObjExpr = path as NodePath<ObjectExpression>
        if (isElasticType(path)) {
          esTypeNodes.push(pathObjExpr)
        }
      }
    }
  }

  traverse(ast, new ElasticTypeCollector())
```

However, I am still curious about computing the distance to the root node. I typed up this to give it a try, but I dont know how to implement `getParent(node)`
```javascript
export function distanceToAncestor(leaf: Node, ancestor: Node): number {
  let edges = 0
  let current = leaf
  while(current !== ancestor) {
    const parent = getParent(current) // how ???
    if (!parent) return -1

    edges++
    current = parent
  }

  return edges
}
```

**Reply by cyrfer** [reply]
Ok I should have tried a little harder. This satisfies that question about calculating the distance to the root node, but I'll try @JLHwung 's idea too.

```javascript
export function parseJson(input: string): ParseResult<File> {
  const ast = parse('(' + input + ')')
  return ast
}

export function getParent(path: NodePath<Node>): NodePath<Node> | null {
  return path.parentPath
}

export function distanceToAncestor(leaf: NodePath<Node>, ancestor: Node): number {
  let edges = 0
  let current = leaf
  while(current.node !== ancestor) {
    const parent = getParent(current)
    if (!parent) return -1

    edges++
    current = parent
  }

  return edges
}

it('should count edges to parent', () => {
  const astRoot = parseJson(JSON.stringify({a: 1, b: {c: 3}}))
  // console.log('astRoot', JSON.stringify(astRoot, null, 2))

  const objExpr: Array< NodePath<ObjectExpression> > = []

  traverse(astRoot, {
    enter(path: NodePath) {
      if (path.isObjectExpression()) {
        objExpr.push(path as NodePath<ObjectExpression>)
      }
    }
  })

  const distances: number[] = objExpr.map(path => {
    return distanceToAncestor(path, astRoot.program)
  })
  console.log('distances', distances)
  // prints
  // distances [ 2, 4 ]
})
```
